docs/tutorial.md

```diff
@@ -737,3 +731,90 @@ Note that even though adding type annotations to .lua files makes it invalid
 Lua, you can still do so and load them from the Lua VM once the Teal package
 loader is installed by calling tl.loader().
 
+
+### Type definitions for third party libraries
+
+You can also create declaration files to annotate the types of third-party Lua
+modules, including C Lua modules. To do so, create a file with the .d.tl
+extension and require it as normal, i.e. `require("minetest")`.
+
+Types defined in this module will will be used as a source of type information 
+checking with `tl check`, even though the real Lua module will be loaded
+instead when requiring the module from Lua or `tl run`.
+
+If the third party library is "globally available", i.e. you do not explicitly
+`require` it, you can inject the types without affecting your generated code
+with a `tlconfig.lua`:
+
+```lua
+return {
+    preload_modules = {
+        "minetest" -- Execute the equivalent of require('minetest') before type-checking the tl script(s)
+    }
+}
+```
+
+#### Visibility
+
+It should be noted that local references in type definition files (or any files
+containing types for that matter) are not available in the requiring module.
+
+This means that constructs like:
+
+```lua
+local MyPointType = record
+    x :number
+    y :number
+end
+
+local MyCompositeType = record
+    center: MyPointType
+end
+
+return {
+    MyPointType = MyPointType,
+    MyCompositeType = MyCompositeType,
+}
+```
+
+will give errors when accessing `MyCompositeType::center`, as `MyPointType` is not
+visible.
+
+There are two ways to define these types:
+
+##### Composite Types
+
+```lua
+local MyCompositeType = record
+    MyPointType = record
+        x :number
+        y :number
+    end
+
+    center :MyPointType
+    -- insert more stuff here
+end
+
+return MyCompositeType
+```
+
+This will mean that references to `MyPointType` must be qualified (or locally declared) as
+`MyCompositeType.MyPointType`.
+
+##### Global Types
+
+While perhaps discouraged in general, for types that are provided as part of the
+execution environment it may be more hygeinic to declare them as global.
```

Comment:
Appreciate the care you took here, but this is a little tricky to give as advice, because the technically the execution environment does not provide these types for real (at least not the type _names_) so ocuppying global names with type definitions is not something to be done lightly. 

Given that exported local and indirectly referenced types work, as described above, I'm inclined we should avoid directing users to global types, at least in the tutorial.

Perhaps instead of this subsection, we could have just a quick note that "Types can be declared with `global`, but in general this should be avoided in reusable modules."

Reply:
If they are injected via preload they shouldn't end up in the generated code at all (unless I am missing something)?

Aye if indirectly referenced types work (I will drag in latest tl tomoz) then 100% agreed that's a nicer way to do it.

Reply:
yeah, not in the generated code, but in the symbol table during compilation. Say I have my own type Bla in my application, and I want to use library foo, and it happens to expose its own global type Bla. Then my own type Bla is shadowing Bla from module foo. If module foo exposed it as foo.Bla, then I could use both with no problems

Reply:
Gotcha. All makes sense now.